### ISSUE TITLE
Dow Jones outperforming ensemble strategy in FinRL_Ensemble_StockTrading_ICAIF_2020 notebook

### ISSUE THREAD

**Issue by AjinkyaBankar** [AI4Finance-Foundation/FinRL#1070]
I replicated the code given in FinRL_Ensemble_StockTrading_ICAIF_2020.ipynb. And to my surprise, Dow Jones outperformed the ensemble strategy. What is the reason?
![Untitled](https://github.com/AI4Finance-Foundation/FinRL/assets/77496039/554da065-e7fc-49f3-aefb-be1059e74fcf)

**Comment by AjinkyaBankar** [comment]
@notBradPitt have you faced such issue?

**Comment by AjinkyaBankar** [comment]
@notBradPitt I am using the same data and same parameters as given in the notebook. My expectation is that I should get the same results as given in their notebook. But, its not happening. Each time I am getting different results.

**Comment by XiaoYangLiu-FinRL** [comment]
> @notBradPitt I am using the same data and same parameters as given in the notebook. My expectation is that I should get the same results as given in their notebook. But, its not happening. Each time I am getting different results.

Please check the randomness behind the codes. In ML, there usually exist multiple sources of randomness.

**Comment by AjinkyaBankar** [comment]
@XiaoYangLiu-FinRL Can you provide more details about sources of randomness for this notebook? And any strategies to fix them?

**Comment by notBradPitt** [comment]
The initial state of the model is initialized randomly. If there's no random seed provided in the code, then I think it'd be impossible to reach the exact same result.

To get a good model, you'll have to finetune the model so it can perform well regardless of the random initial state or data

**Comment by mmmarchetti** [comment]
> @XiaoYangLiu-FinRL Can you provide more details about sources of randomness for this notebook? And any strategies to fix them?

As @notBradPitt pointed out, the initial state of the model can be initialized randomly. This randomness can lead to inconsistent results across different runs, especially if no random seed is provided in the code.

To get consistent results, one way is to control this randomness by setting a fixed seed value. By doing this, the random processes will always produce the same result, making our results reproducible across different runs. Here's how I managed to fix this:

```
agent = DRLAgent(env=env_train)
model_a2c = agent.get_model(model_name="a2c", 
                                                  model_kwargs=params, 
                                                  seed=1)
```

By setting **seed=1**, the randomness behind the model's initial state is controlled, ensuring that you get the same result every time you run the notebook, given that all other parameters and data remain the same.

Of course, setting a seed is just one step. To achieve maximum profit, you might still need to fine-tune other hyperparameters and ensure the model doesn't underfit or overfit the data.

Hope this helps!

**Comment by AjinkyaBankar** [comment]
@notBradPitt @mmmarchetti It can be a good way to replicate the results by choosing a seed value. But won't this lead to limited design space for model training?

**Comment by mmmarchetti** [comment]
@AjinkyaBankar  You raise an important point. Setting a seed value does indeed make the results reproducible, but it can also limit the randomness, which in turn can affect the exploration of the model's design space during training. Let's break this down:

Reproducibility: When we set a seed, it ensures that every time we run the code, the initial state of the model, the random splits of data, or any other random processes will be the same. This is crucial when we want to share our results with others or when we want to compare different strategies, as it guarantees that the variations in results are only due to the changes in the strategies and not because of inherent randomness.

Design Space Exploration: In the context of machine learning, especially reinforcement learning, exploration is key. Agents often need to explore various strategies to find the optimal one. If we fix the seed and thereby limit randomness, there's a possibility that the model might not explore certain paths or strategies that it might have, had the randomness been different.

However, it's essential to strike a balance. While a fixed seed aids in reproducibility, it might be beneficial to run the model with different seeds to understand the variability in the model's performance. This way, you can ensure that the model is robust and not just optimized for a specific initial state.

In summary, while setting a seed is great for reproducibility, it's equally important to periodically change the seed and evaluate the model's performance to ensure its robustness across different initial states.

**Comment by Mr-Nobody-dey** [comment]
Please look at this term in the code "df_actions_td3, here td3 is the model name, so for a2c model, it will be a2c". Write that dataframe to an excel sheet and then investigate the actions of the RL agent during the trading period.

**### What I have found in my investigation is:**
1. The agents bought major number of stocks on the very fast date of trading and hold on to them throughout the trading period. The agents did not sell any stock ( even when their were sharp fall in stock price) throughout the trading period although it bought very few stocks in between.

2. Thus the performance of the agent largely depended on the performance of those stocks.
3. If those selected stocks perform better than index, it will show a good return in the final performance figure and vice versa.

**### Solution:**
I don't know yet but i guess:
1. Longer period of training
2. Adding more indicators
3. Changing the reward from investment turnover to other parameters (such as based on the right timing of entry and exit) might help.

### **Please comment back if you also had similar findings.**
